src/vm/gchelpers.cpp

```diff
@@ -950,21 +962,20 @@ OBJECTREF   AllocateObjectArray(DWORD cElements, TypeHandle elementType, BOOL bA
     _ASSERTE(arrayType.GetInternalCorElementType() == ELEMENT_TYPE_SZARRAY);
 #endif //_DEBUG
 
-    return AllocateArrayEx(ClassLoader::LoadArrayTypeThrowing(elementType),
-                           (INT32 *)(&cElements),
-                           1,
-                           bAllocateInLargeHeap);
+    return AllocateSzArray(ClassLoader::LoadArrayTypeThrowing(elementType), (INT32) cElements, GC_ALLOC_NO_FLAGS, bAllocateInLargeHeap);
 }
 
-
-STRINGREF SlowAllocateString( DWORD cchStringLength )
+STRINGREF AllocateString( DWORD cchStringLength )
 {
     CONTRACTL {
         THROWS;
         GC_TRIGGERS;
         MODE_COOPERATIVE; // returns an objref without pinning it => cooperative
     } CONTRACTL_END;
 
+#if defined(_TARGET_X86_)
+    return STRINGREF(HCCALL1(fastStringAllocator, cchStringLength));
```

Comment:
Can this end up in infinite recursion? `fastStringAllocator` -> `UnframedAllocateString` -> `AllocateString` -> `fastStringAllocator`

Reply:
These x86-specific codepaths look like a bug farm. It would be nice to get rid of them.

Reply:
it looks like it could get recursive. will have to fix.

---
In reply to: [279989632](https://github.com/dotnet/coreclr/pull/24096#discussion_r279989632) [](ancestors = 279989632)

Reply:
I do not know the history of this code. It looks complex and I assume there must be perf benefits from this (or was at some point). 
I will look closer at what is being achieved in that code and whether there is a chance of regressing something.

---
In reply to: [279989761](https://github.com/dotnet/coreclr/pull/24096#discussion_r279989761) [](ancestors = 279989761)

Reply:
There was a benefit from this in 1999 when this was written and the calling patterns were completely different. The fact that this optimization is not present on x64 that is our peak-performance target suggests that it is not important.